cli/exp_rpty.go

```diff
@@ -0,0 +1,217 @@
+	}
+	ws, agt, err := getWorkspaceAndAgent(ctx, inv, client, true, args.NamedWorkspace)
+	if err != nil {
+		return err
+	}
+
+	var ctID string
+	if args.Container != "" {
+		cts, err := client.WorkspaceAgentListContainers(ctx, agt.ID, nil)
+		if err != nil {
+			return err
+		}
+		for _, ct := range cts.Containers {
+			if ct.FriendlyName == args.Container || ct.ID == args.Container {
+				ctID = ct.ID
+				break
+			}
+		}
+		if ctID == "" {
+			return xerrors.Errorf("container %q not found", args.Container)
+		}
+	}
+
+	if err := cliui.Agent(ctx, inv.Stderr, agt.ID, cliui.AgentOptions{
+		FetchInterval: 0,
+		Fetch:         client.WorkspaceAgent,
+		FetchLogs:     client.WorkspaceAgentLogsAfter,
+		Wait:          false,
+	}); err != nil {
+		return err
+	}
+
+	// Get the width and height of the terminal.
+	var termWidth, termHeight uint16
+	stdoutFile, validOut := inv.Stdout.(*os.File)
+	if validOut && isatty.IsTerminal(stdoutFile.Fd()) {
+		w, h, err := term.GetSize(int(stdoutFile.Fd()))
+		if err == nil {
+			//nolint: gosec
+			termWidth, termHeight = uint16(w), uint16(h)
+		}
+	}
+
+	// Set stdin to raw mode so that control characters work.
+	stdinFile, validIn := inv.Stdin.(*os.File)
+	if validIn && isatty.IsTerminal(stdinFile.Fd()) {
+		inState, err := pty.MakeInputRaw(stdinFile.Fd())
+		if err != nil {
+			return xerrors.Errorf("failed to set input terminal to raw mode: %w", err)
+		}
+		defer func() {
+			_ = pty.RestoreTerminal(stdinFile.Fd(), inState)
+		}()
+	}
+
+	conn, err := workspacesdk.New(client).AgentReconnectingPTY(ctx, workspacesdk.WorkspaceAgentReconnectingPTYOpts{
+		AgentID:       agt.ID,
+		Reconnect:     reconnectID,
+		Command:       strings.Join(args.Command, " "),
+		Container:     ctID,
+		ContainerUser: args.ContainerUser,
+		Width:         termWidth,
+		Height:        termHeight,
+	})
+	if err != nil {
+		return xerrors.Errorf("open reconnecting PTY: %w", err)
+	}
+	defer conn.Close()
+
+	cliui.Infof(inv.Stderr, "Connected to %s (agent id: %s)", args.NamedWorkspace, agt.ID)
+	closeUsage := client.UpdateWorkspaceUsageWithBodyContext(ctx, ws.ID, codersdk.PostWorkspaceUsageRequest{
+		AgentID: agt.ID,
+		AppName: codersdk.UsageAppNameReconnectingPty,
+	})
+	defer closeUsage()
+
+	stdinDone := make(chan struct{})
+	stdoutDone := make(chan struct{})
+	stderrDone := make(chan struct{})
+	done := make(chan struct{})
+
+	go func() {
+		defer close(stdinDone)
+		// This is how we send commands to the agent.
+		br := bufio.NewScanner(inv.Stdin)
+		// Split on bytes, otherwise you have to send a newline to flush the buffer.
+		br.Split(bufio.ScanBytes)
+		je := json.NewEncoder(conn)
+		for br.Scan() {
+			if err := je.Encode(map[string]string{
+				"data": br.Text(),
+			}); err != nil {
+				return
+			}
+		}
+	}()
+	go func() {
+		defer func() {
+			close(stdoutDone)
+		}()
+		_, _ = io.Copy(inv.Stdout, conn)
+	}()
+	go func() {
+		defer func() {
+			close(stderrDone)
+		}()
+		_, _ = io.Copy(inv.Stderr, conn)
+	}()
```

Comment:
There's no separate stderr/stdout on conn, so this will send part of the messages to stdout and other part to stderr, we can simply remove this.

Reply:
That's a really useful observation!